### ISSUE TITLE
Offline smoothing breaks with states with more than one dimension

### ISSUE THREAD

**Issue by jorgemcgomes** [nchopin/particles#3]
I just started using this library, and I'm no expert in PFs, so excuse me if I'm making any wrong assumption. But it seems I've found a bug, where offline smoothing breaks with states with more than one dimension.
This happens with backward_sampling linear, ON2, as well as backward_sampling_qmc.

When the state has multiple variables (i.e., IndepProd distribution), trying to run smoothing fails with the following error:

```
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
<ipython-input-242-3ee6f0106081> in <module>
----> 1 smooth_trajectories = pf.hist.backward_sampling(100)

venv/lib/python3.7/site-packages/particles/smoothing.py in backward_sampling(self, M, linear_cost, return_ar)
    177             ar = self._backward_sampling_ON(M, idx)
    178         else:
--> 179             self._backward_sampling_ON2(M, idx)
    180         # When M=1, we want a list of states, not a list of arrays containing
    181         # one state

venv/lib/python3.7/site-packages/particles/smoothing.py in _backward_sampling_ON2(self, M, idx)
    221             for t in reversed(range(self.T - 1)):
    222                 lwm = (self.wgt[t].lw + self.model.logpt(t + 1, self.X[t],
--> 223                                                      self.X[t + 1][idx[t + 1, m]]))
    224                 idx[t, m] = rs.multinomial_once(rs.exp_and_normalise(lwm))
    225 

venv/lib/python3.7/site-packages/particles/state_space_models.py in logpt(self, t, xp, x)
    342     def logpt(self, t, xp, x):
    343         """PDF of X_t|X_{t-1}=xp"""
--> 344         return self.ssm.PX(t, xp).logpdf(x)
    345 
    346     def upper_bound_trans(self, t):

venv/lib/python3.7/site-packages/particles/distributions.py in logpdf(self, x)
    834 
    835     def logpdf(self, x):
--> 836         return sum([d.logpdf(x[:, i]) for i, d in enumerate(self.dists)])
    837 
    838     def ppf(self, u):

venv/lib/python3.7/site-packages/particles/distributions.py in <listcomp>(.0)
    834 
    835     def logpdf(self, x):
--> 836         return sum([d.logpdf(x[:, i]) for i, d in enumerate(self.dists)])
    837 
    838     def ppf(self, u):

IndexError: too many indices for array
```

It appears that the logpdf function is expecting an array of shape (1, d) in the last argument (x), but an array of shape (d,) is being passed to it (d being number of state variables).

I tried modifying the `_backward_sampling_ON2` function by adding a reshape, as such:
Where it is currently:
`lwm = (self.wgt[t].lw + self.model.logpt(t + 1, self.X[t], self.X[t + 1][idx[t + 1, m]]))`
I replaced with:
`lwm = (self.wgt[t].lw + self.model.logpt(t + 1, self.X[t], self.X[t + 1][idx[t + 1, m]].reshape(1, -1)))`

This simple fix seems to have fixed it, and the function doesn't crash anymore.

I imagine this "fix" will create problems in other situations, so it's not really a fix, but this does suggest that there is some problem here.

**Comment by jorgemcgomes** [comment]
An update: turns out this was also failing with the ON2 online smoothing.

I was able to get a better and more universal fix by modifying the IndepProd instead, and leaving all the rest as it is:
```
class FixedIndepProd(IndepProd):

    def logpdf(self, x):
        assert x.ndim in (1, 2)
        if x.ndim == 1:
            x = x.reshape(1, -1)
        return super().logpdf(x)

    def ppf(self, u):
        assert u.ndim in (1, 2)
        if u.ndim == 1:
            u = u.reshape(1, -1)
        return super().ppf(u)
```

**Comment by nchopin** [comment]
Hi,
thanks for using particles and reporting this! 
Could you plz also post the code that generates the error? I am interested in particular in how you use IndepProd when you define your state-space model. 
Thanks!

**Comment by jorgemcgomes** [comment]
Hi,

Thank you for this very nice and well documented library.

Here's a notebook that replicates the bug and the fix, and the data needed to run it:
[particles_test.zip](https://github.com/nchopin/particles/files/3800783/particles_test.zip)

**Comment by nchopin** [comment]
Bug **confirmed**... Drat, this is quite serious one. I am very busy these days (teaching) but will try to fix this asap.

**Comment by nchopin** [comment]
I think I fixed the bug in experimental branch. Test case submitted by @jorgemcgomes now runs fine. Some extra comments for @jorgemcgomes:

a. I adapted a bit the fix you recommended, and used the ellispis instead: 

class IndepProd
    # ... 
    def logpdf(self, x):
        return sum([d.logpdf(x[..., i]) for i, d in enumerate(self.dists)])

In this way, IndepProd now accepts for either a (N, d) array, or a (d) array. 

b. you mentioned also online smoothing, but did you implement method add_func, as explained here? 
https://particles-sequential-monte-carlo-in-python.readthedocs.io/en/latest/collectors.html#on-line-smoothing

c. please note that backward smoothing does not *really work* for models involving Dirac masses (or any model such that X_t given X_{t-1} does not have a proper probability density). Backward sampling tries to reconstruct smoothing trajectories, by trying to connect/pair particles X_t^n with previous particles X_{t-1}^m, etc.  But in the model you consider, the probability of pairing will be 0 for all the ancestors X_{t-1}^m excepts the one that generated X_t^n (because of the Dirac masses). So the only trajectories you will obtain will be exactly those you can obtain from the genealogy of the particle system (without doing the backward pass), by invoking:

forward_trajectories = pf.hist.compute_trajectories()

I am actually surprised that my implementation works even in this degenerate case, and I would recommend instead the method above, which simply computes the forward trajectories. 

Please let me know if there is anything unclear of if something still does not work as expected. And thanks again for reporting this issue.

**Comment by jorgemcgomes** [comment]
Thanks for the clarification on smoothing.
I've been trying to learn a bit more about the smoothing techniques, but to be honest, the math in the papers is sometimes a bit too much for my math skills.

b) I was trying with a very naive add_func that didn't really do anything useful: just Xp - X
I have some trouble understanding what add_func is supposed to do, and I couldn't find any meaningful examples for it. Couldn't find any relevant literature for it either, so I gave up on that.
Instead, I implemented fixed-lag smoothing in the lib, which suits my purposes just fine and is working well. (based on this [paper](http://ftp.phmsociety.org/sites/phmsociety.org/files/phm_submission/2009/phmc_09_37.pdf), section 2.2)

c) I guess that explains why I wasn't getting very noticeable improvements. But the results were definitely usable (slightly better than what I got with the weighted average moments).
In any case, the computational performance was not good enough for my purposes anyway, so I also gave up on that.

On a different note, something else that I had to "hack" in order to get it to work with my problem:
- support for irregular time steps (dt is different in each update)
- processing of streaming data

I'm not yet sure if I'll implement some general solution in the library to support this, or if I'll just hack around it. But if I do get to a point where I have something that looks good, I'll make a pull request (together with the fixed-lag smoothing), and then you can decide what to do with it.

On yet another note, I haven't tried this, but it seems to me that the APF implementation is broken.
Shouldn't the logeta function take as arguments t, X, and data?
The documentation in StateSpaceModel says so:
 `logeta(self, t, x, data)`: the auxiliary weight function at time t`

and there's also an example:
```
class StochVol(StateSpaceModel):
    def logeta(self, t, x, data):
```

But if you look at AuxiliaryPF and AuxiliaryBootstrap, you see that it's missing the x:
```
    def logeta(self, t):
        return self.ssm.logeta(t, self.data)
```

Sorry, I got a bit carried away here, maybe these problems should belong on their own issues.

**Comment by nchopin** [comment]
Hi,
* logetat: good catch. I started a new issue (#3) and implemented a fix in branch experimental.
* fix-lag smoothing: see issue #4, this is an enhancement I was considering any way, it should not be too much work. 
* add_func and the rest: I will PM you shortly. 

This thread should be automatically closed when I merge experimental -> master. I should do so in the coming days.
